documentation/DEVELOPER-GUIDELINES.md

```diff
@@ -23,13 +23,12 @@ In general, the [Rust API Guidelines](https://rust-lang.github.io/api-guidelines
 
 ## Construction and Destruction of Drivers
 
-- Drivers must take peripherals via the `PeripheralRef` pattern - they don't consume peripherals directly.
-- If a driver requires pins, those pins should be configured using `fn with_signal_name(self, pin: impl Peripheral<P = impl PeripheralInput> + 'd) -> Self` or `fn with_signal_name(self, pin: impl Peripheral<P = impl PeripheralOutput> + 'd) -> Self`
+- If a driver requires pins, those pins should be configured using `fn with_signal_name(self, pin: impl PeripheralInput<'d>) -> Self` or `fn with_signal_name(self, pin: PeripheralOutput<'d>) -> Self`
   - The `with_signal_name` functions should be placed in an `impl` block that is generic over the driver mode (i.e. they shouldn't be constrained to just `Blocking` or just `Async`) unless there is a technical reason to do otherwise.
 - If a driver supports multiple peripheral instances (for example, I2C0 is one such instance):
   - The driver should not be generic over the peripheral instance.
   - The author must to use `crate::any_peripheral` to define the "any" peripheral instance type.
-  - The driver must implement a `new` constructor that automatically converts the peripheral instance into the any type.
+  - The driver must implement a `new` constructor that automatically converts the peripheral instance into the any type. This is usually written as `fn new(peri: impl Instance + Into<AnyInstance<'d>>)`.
```

Comment:
I am a bit torn here, `impl Instance + Into<AnyInstance<'d>>` isn't great. We could add the conversion to the `Instance` trait, but it would then have a lifetime - this was done for GPIO peripheral signals, see `PeripheralInput<'d>`, so I guess we could go that way, although I'm not quite sure about long-term implications. According to Dario traits + lifetimes are a fast track to type system hell.

Reply:
> add the conversion to the Instance trait, but it would then have a lifetime

This also means we have to mandate lifetimes for all peripheral Instance traits, or making sure we don't have an Instance trait for peripherals that only have a single option currently. If we have a lifetimeless Instance, and we need to add support for a chip that has multiple peripherals of that function, we'll be in trouble.

Reply:
I guess "do nothing (as in, keep using `impl Instance + Into<AnyInstance<'d>>`) and stay inconsistent with PeripheralOutput" is an option, too.